src/seraphis_wallet/transaction_history.h

```diff
@@ -0,0 +1,195 @@
+//    conditions and the following disclaimer.
+//
+// 2. Redistributions in binary form must reproduce the above copyright notice, this list
+//    of conditions and the following disclaimer in the documentation and/or other
+//    materials provided with the distribution.
+//
+// 3. Neither the name of the copyright holder nor the names of its contributors may be
+//    used to endorse or promote products derived from this software without specific
+//    prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+// MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
+// THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+// STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+// THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+// The goal of this component (SpTransactionHistory) is to store all the relevant information (SpTransactionStoreV1)
+// from transactions made by the wallet and allow the user to efficiently access these information later (to visualize
+// specific info from enotes or generate knowledge proofs for example).
+
+#pragma once
+
+// local headers
+#include "crypto/crypto.h"
+#include "ringct/rctTypes.h"
+#include "seraphis_core/jamtis_destination.h"
+#include "seraphis_core/jamtis_payment_proposal.h"
+#include "seraphis_impl/enote_store.h"
+#include "seraphis_main/contextual_enote_record_types.h"
+#include "seraphis_main/txtype_squashed_v1.h"
+
+// third party headers
+#include <boost/range.hpp>
+#include "boost/range/iterator_range.hpp"
+
+// standard headers
+#include <cstddef>
+#include <cstdint>
+#include <functional>
+#include <map>
+#include <tuple>
+#include <unordered_map>
+#include <unordered_set>
+#include <vector>
+
+// forward declarations
+
+using namespace sp::jamtis;
+using namespace sp;
+
+// struct to store the relevant info from a tx
+struct TransactionRecordV1
+{
+    // key images of spent enotes for tracking purposes
+    std::vector<crypto::key_image> legacy_spent_enotes;
+    std::vector<crypto::key_image> sp_spent_enotes;
+
+    // JamtisPaymentProposals
+    std::vector<JamtisPaymentProposalSelfSendV1> selfsend_payments;
+    std::vector<JamtisPaymentProposalV1> normal_payments;
+
+    // fees and total sent
+    // useful to store here also instead of looking directly at the enotes and blockchain
+    rct::xmr_amount amount_sent;
+    rct::xmr_amount fee_sent;
+};
+
+// struct to efficiently store/sort records
+struct SpTransactionStoreV1
```

Comment:
If I come around I will write a separate issue with details about the question whether in the Seraphis wallet code we should follow the Seraphis library convention of versioning class names, like here with `SpTransactionStoreV1` and not simply `TransactionStore`.

In short: I think in the Seraphis library this versioning is a good idea, because beside `JamtisPaymentProposalV1` there may be soon `JamtisPaymentProposalV2` and even `JamtisPaymentProposalV3`, and the library will have to support them all, concurrently. You won't be able to give up `JamtisPaymentProposalV1`. Indeed, if you could, the whole versioning would make no sense IMHO.

Is the same true about this transaction store? Is there a good probability that soon we will have a `SpTransactionStoreV2` and that, even more importantly, wallet code will have to juggle with a `SpTransactionStoreV1` **and** a `SpTransactionStoreV2` object in parallel for some hard technical reason?

That somehow sounds pretty improbable to me. That leads me to a proposal to drop the "V1" here and go for `SpTransactionStore`.

But as I said, maybe that decision needs a broader discussion.

Reply:
https://github.com/seraphis-migration/wallet3/issues/62